### DISCUSSION TITLE
[Feature request] Exile Leveling by HeartofPhos

### DISCUSSION THREAD

**Discussion by ghost** [Lailloken/Exile-UI#88] [Ideas & Feedback]
It would be great if https://github.com/HeartofPhos/exile-leveling can be added into LUI (with the dev's permission ofc). Like an overlay that shows the next steps dynamically by scanning the current zone name or something.

**Comment by Lailloken** [comment]
Hmm, interesting. I saw the reddit post this morning and thought this was really neat. I can see a way of using both together by exporting their guide as a bullet-point list and importing it into my notepad feature and creating a grouped widget -- the text in the guide cannot be copied at the moment. I could implement zone tracking, so that it automatically changes the pages in the widget whenever you reach the zone mentioned in the current step in the guide. I could also implement scrolling to the next page by omni-clicking the quest-items and gems mentioned in the text.

The thing is: This would put LLK-UI in the foreground and kind of pull attention away from their work. The user would only use their work to generate the guide and export it, then use LLK-UI to make use of it. I guess you could propose it to them and see what they think. It wouldn't be much work on my side and I think it would be a useful implementation.

EDIT: Okay, I tried to have a look at what's behind the project (code-wise) and I have no idea what's going on or how it works, and the documentation is basically 0 for someone like me. So if any form of integration were to happen, it would have to be by copying/exporting the guide as I described.

**Reply by HeartofPhos** [reply]
Happy to help with integrating if there is any demand, shouldn't be too hard export the guide into a format more suitable for LUI

**Reply by Lailloken** [reply]
> Happy to help with integrating if there is any demand, shouldn't be too hard export the guide into a format more suitable for LUI

Cool, thanks a lot. A simple 'export' button that copies the guide into the clipboard would already suffice (since you showed me the formatting and templates). I would implement a system to parse the text and adjust it accordingly for presentation in an AHK GUI. Now that I have a better overview of things, I would actually make this its own feature instead of using the [notepad & text-widgets](https://github.com/Lailloken/Lailloken-UI/wiki/Notepad-&-Text-widgets) feature: It would have its own section in the settings menu with a link to your GitHub page, a direct link to the tool to generate a guide, and an import button.

I'm always cautious when it comes to diverting traffic/usage away from other creators, so I would also be okay with importing and displaying only one act at a time (i.e. usage between our tools wouldn't be as one-sided). It would impact the user-experience, but I don't think the target audience of leveling guides would play fast enough to be bothered by that. Let me know what you think.

**Reply by HeartofPhos** [reply]
I think it depends on the user flow you want in LUI, the tenplates don't contain gem information, I add that based on quests at runtime

If you want to add UI for parsing a pob & templates in AHK and simply want the raw tenplate files it's probably simpler to just download directly from Github that way you can anchor to a commit and any futures updates I may make won't break your code

Alternatively I can add an export button specifically for LUI in JSON with all the relevant data, in that case I'd imagine a flow like:
- user imports their POB on the website
- updates gems as required
- hit Export to LUI to copy entire route + gem information to clipboard
- paste into LUI and go from there

Either option is fine by me, I don't mind traffic being diverted at all, as long as there's attribution somewhere users can decide which experience they prefer

**Reply by Lailloken** [reply]
I was thinking of the alternative way you just described because I'd want people to still use your tool as the source. That's generally how I want to handle feature-additions to my tool -- i.e. if there's an established tool or feature that does something I find interesting and cannot create my own version/take of, I try to come up with a way to implement a quick-access from within the client in order to eliminate alt-tabbing (and searching manually) and to improve user-experience. For example, there's this context-menu for items that gives the user quick access to features within my tool, or to popular 3rd-party web sources (e.g. community wiki, poedb crafting resources, vorici calculator, etc.).

And that's also what I had in mind for your guide generator: The settings menu where the eventual overlay will be configured will have a link to access your tool, and an import-button to load the generated guide into the dynamic overlay. Having guides in the client as an overlay was the main reason I implemented text-widgets, only that the user had to format them themselves. Your tool is the perfect framework to streamline the process, and also to make the overlays dynamic because I can build around your specific guide-format when reading the client.txt.

I referred to the templates because I expected the export to have the same structure. I don't mind if it will look different, as long as it has a fixed structure -- whatever is the easiest to implement from your end, I'll just find a way to adapt to it. And you don't have to specifically label the export feature with any reference to my tool because it may also find more general use (e.g. LiveSplit or other overlay tools) and because I personally wouldn't like the connotation/implication. It's my tool that taps into your project which is independent and fully-functional on its own -- there just happens to be demand for a use-case you're not intending to pursue for now.

**Reply by HeartofPhos** [reply]
Ok I've given it a first pass
![image](https://user-images.githubusercontent.com/43216176/182931309-10cbb566-71fc-46c2-8d6b-e80ebbd26264.png)

```ts
type Root = Route[];
type Route = Step[];
type Step = FragmentStep | GemStep;

interface GemStep {
  type: "gem_step";
  requiredGem: RequiredGem;
  rewardType: "quest" | "vendor";
}

interface RequiredGem {
  id: string;
  uid: number;
  note: string;
}

interface FragmentStep {
  type: "fragment_step";
  parts: (string | Fragment)[];
}

type Fragment =
  | KillFragment
  | ArenaFragment
  | AreaFragment
  | EnterFragment
  | LogoutFragment
  | WaypointFragment
  | GetWaypointFragment
  | PortalFragment
  | QuestFragment
  | QuestTextFragment
  | GenericFragment
  | TrialFragment
  | AscendFragment
  | DirectionFragment
  | CraftingFragment;

interface KillFragment {
  type: "kill";
  value: string;
}

// Sub areas that aren't defined in the game data
interface ArenaFragment {
  type: "arena";
  value: string;
}

interface AreaFragment {
  type: "area";
  areaId: Area["id"];
}

interface LogoutFragment {
  type: "logout";
  areaId: Area["id"];
}

interface EnterFragment {
  type: "enter";
  areaId: Area["id"];
}

interface WaypointFragment {
  type: "waypoint";
  areaId: Area["id"] | null;
}

interface GetWaypointFragment {
  type: "get_waypoint";
}

interface GenericFragment {
  type: "generic";
  value: string;
}

interface PortalFragment {
  type: "portal";
  targetAreaId?: Area["id"];
}

interface CraftingFragment {
  type: "crafting";
  crafting_recipes: string[];
}

// 0-7 cardinal directions, clockwise starting at "Up" 
interface DirectionFragment {
  type: "dir";
  dirIndex: number;
}

interface QuestFragment {
  type: "quest";
  questId: Quest["id"];
  rewardOffers: number[];
}

interface QuestTextFragment {
  type: "quest_text";
  value: string;
}

interface TrialFragment {
  type: "trial";
}

interface AscendFragment {
  type: "ascend";
  version: "normal" | "cruel" | "merciless" | "eternal";
}
```
The format is basically exactly as I used it to render the UI so it uses gem, area and quest ids that I get from data mining, json blobs for that data is stored here https://github.com/HeartofPhos/exile-leveling/tree/main/common/data/json

Depending on your needs I can export actual values e.g. quest/area/gem names so you don't need to have the exported data available for referencing, but this is at least a starting point and we can narrow down the scope as you see fit

I'm not currently exporting any of the progress information but that can be added if it makes sense.

Let me know if you'd like any changes or clarification

**Reply by Lailloken** [reply]
> I'm not currently exporting any of the progress information but that can be added if it makes sense.

Do you mean what has been ticked off the list/route already? If so, I personally won't use it because I won't be able to. This is a mockup of what I have in mind, using my text-widgets to illustrate:

![image](https://user-images.githubusercontent.com/61888437/182937515-0681fc6c-e61e-4193-9046-1b47a4de51da.png)

I'll probably have to do some restructuring internally and create my own progression system because some steps are not trackable by reading the client.txt (e.g. handing in quests that don't involve a quest-item, unlocking crafting recipes, etc.), so the dynamic overlay wouldn't know when to advance. Some steps will need to be grouped, and I'll have to hand-pick these 'milestones' that advance the overlay in order to minimize required user-input, while also making sure I don't group too many together at once.

> Depending on your needs I can export actual values e.g. quest/area/gem names so you don't need to have the exported data available for referencing

That would be great if it's not too much work. Basically, the closer you can get it to what's displayed in the route, the more understandable/usable it will be for me (I know basically nothing apart from AHK, and what I know is a product of self-teaching and tinkering). Looking at the export right now, my gut-feeling tells me I can work things out with a bit of playing around once it contains the actual values. Would it also be possible to change the formatting a bit? Currently, it exports everything as one continuous string.

Thanks a lot for your work, much appreciated.

**Reply by HeartofPhos** [reply]
> Do you mean what has been ticked off the list/route already?

Yep if you don't need/can't use it then I'll leave it out.

> Would it also be possible to change the formatting a bit? Currently, it exports everything as one continuous string

Currently it's exporting minified JSON, you can expand it using any JSON formatter e.g. https://jsonformatter.org/
My AHK knowledge is very limited but it doesn't seem to have native JSON support from what I can tell, all I can find is https://github.com/cocobelgica/AutoHotkey-JSON
Is there another data format you'd prefer it to be exported in?

I don't mind exporting human-readable data instead of Ids where appropriate, if you take a look at the JSON data e.g. https://github.com/HeartofPhos/exile-leveling/blob/main/common/data/json/areas.json and let me know which fields you're interested in I can surface them.

I'm hesitant to do that blindly since I may exclude data you'd find useful, for example I know the client.txt uses `area.id` when it's generating an instance which is the only way I can see differentiating different zones with identical names between acts, but there's likely many other things I lack context, can narrow it down once you know what data you're interested in.

**Reply by Lailloken** [reply]
> Is there another data format you'd prefer it to be exported in?

Now that I know what I'm looking at, it makes more sense to learn how to handle this kind of data (I've seen JSON thrown around here and there, but never knew what to make of it because I never had to interact with it). I'll get back to you if that fails miserably.

> I know the client.txt uses `area.id` when it's generating an instance which is the only way I can see differentiating different zones with identical names between acts

Oh right, I have not noticed that. This is the first time I'm working more in-depth with the client.txt (I'm persuing this user-request to get some experience handling client-generated data). I was just banking on the fact that it's highly unlikely a player would disrupt the flow of the dynamic overlay by randomly backtracking to an earlier story-part at exactly the right time, so I was just going to use the txt's verbose output. Thanks for pointing that out, one more reason to figure out that JSON stuff (reading it cracks me up every time because I have to think of Heavy Rain).

**Reply by Lailloken** [reply]
I made some progress today -- the import feature is able to create the full route for act 1 from the exported data:

01. Find and kill Hillock
02. head to areaId_1_1_town
03. Hand in Enemy at the Gate
04. buy Arcane Surge Support
05. head to areaId_1_1_2
06. activate waypoint
07. head to areaId_1_1_3
08. Find the three glyphs
09. head to areaId_1_1_4_1
10. waypoint-travel to areaId_1_1_2
11. head to areaId_1_1_2a
12. Find and kill Hailrake, take Medicine Chest
13. exit to character selection, enter areaId_1_1_town
14. Hand in Breaking Some Eggs
15. take reward: Orb of Storms
16. Hand in Mercy Mission
17. waypoint-travel to areaId_1_1_4_1
18. Find bridge, place portal
19. head to areaId_1_1_5
20. head to areaId_1_1_6
21. head to areaId_1_1_7_1
22. waypoint-travel to areaId_1_1_town
23. take portal to targetAreaId_1_1_4_1
24. go right of bridge, then head to areaId_1_1_4_0
25. Find and kill The Dweller of the Deep
26. exit to character selection, enter areaId_1_1_town
27. Hand in The Dweller of the Deep
28. Hand in The Caged Brute
29. take reward: Added Lightning Damage Support
30. waypoint-travel to areaId_1_1_7_1
31. Complete the lab-trial
32. head to areaId_1_1_7_2
33. head to The Warden's Quarters, kill Brutus
34. exit to character selection, enter areaId_1_1_town
35. Hand in The Caged Brute
36. take reward: Flame Dash
37. buy Clarity
38. waypoint-travel to areaId_1_1_8
39. Go down the ledge next to the road
40. head to areaId_1_1_9
41. activate waypoint
42. Find areaId_1_1_9a, place portal and move along
43. head to areaId_1_1_11_1
44. waypoint-travel to areaId_1_1_town
45. take portal to targetAreaId_1_1_9
46. head to areaId_1_1_9a
47. Find Allflame
48. exit to character selection, enter areaId_1_1_town
49. waypoint-travel to areaId_1_1_9
50. Find and kill Captain Fairgraves
51. exit to character selection, enter areaId_1_1_town
52. Hand in The Marooned Mariner
53. Hand in The Siren's Cadence
54. take reward: Arc
55. waypoint-travel to areaId_1_1_11_1
56. head to areaId_1_1_11_2
57. head to Merveil's Lair, kill Merveil

The area-IDs will be resolved at a later stage when some of the individual steps will be grouped. I have a question regarding parsing the data: Is it normal for data to be accessed alphabetically (in a for-loop), or is it an AHK thing? Each step in the export data has 'type' and 'parts'/'requiredGems'/etc., and these are accessed alphabetically on my end, instead of the order in the original string, which means I have to do manual tweaking.

For example, step 10 in act one is "waypoint -> the coast", but letting the script parse the data without tweaks will read the area-ID first, then the waypoint. It wouldn't surprise me if I'm missing something here and being dumb about it, so could you enlighten me?

**Reply by HeartofPhos** [reply]
I suspect you may be over complicating how you access the data
```ahk
#Include ./JSON.ahk

; Exported Routes
FileRead, routes_str, ./routes.json
routes := JSON.Load(routes_str)

; https://github.com/HeartofPhos/exile-leveling/blob/main/common/data/json/areas.json
FileRead, areas_str, ./areas.json
areas := JSON.Load(areas_str)

ParseStep(step)
{
    global areas
    ; if(step.["type"] == "fragment_step") ; Can access fields like this too
    if(step.type == "fragment_step")
    {
        output := ""
        Loop % step.parts.Length()
        {
            part = % step.parts[A_Index]
            if (part is string)
            {
                ; Plain Text
                output := output part
            }
            else
            {
                ; Fragment
                Switch % part.type
                {
                    Case "enter":
                        output := output areas[part.areaId].name
                    Case "waypoint":
                        output := output "Waypoint to " areas[part.areaId].name
                    ; etc, etc
                }
            }
        }

        MsgBox % output
    }
}
; Route 1, Step 8
ParseStep(routes[1][8])
; Route 1, Step 9
ParseStep(routes[1][9])
```
This is roughly the same flow I use when rendering the web version from this same data, little bit scuffed because message box doesn't render "➞" properly but should work fine with proper UI

That said this is obviously a simplified example, some fragments like `waypoint` have optional properties e.g. areaId can be empty and on the website i basically of
```
if (areaId == null)
    render waypoint.png "Waypoint"
else
    render waypoint.png "Waypoint ➞" areas[areaId]
```

But you can can see the full list of fields in my earlier comment

**Reply by Lailloken** [reply]
> I suspect you may be over complicating

That's my specialty, and being stubborn about figuring stuff out myself. At least I can say I got it somewhat working by intuition ![Untitled 1](https://user-images.githubusercontent.com/61888437/183258117-f0e84c7b-9471-4fea-b79a-aa9312276a7c.jpg).

I gotta be honest, working on this has not been fun. I hope the stuff I learn makes up for that. Thanks for helping me out. I might go silent-mode for a while.

**Reply by HeartofPhos** [reply]
> That's my specialty, and being stubborn about figuring stuff out myself. At least I can say I got it somewhat working by intuition

Welcome to me struggling for 10min trying to figure out how to get the length of an array in AHK, pain of unfamiliar tech

Feel free to pull me in if you setup a branch and need some extra hands/eyes

**Reply by Lailloken** [reply]
Alright, I used Sunday to collect my thoughts a bit and found out that I'm actually getting quite irritated by this. I usually don't take feature requests if I don't see myself, or many people in general, using it -- unless I get something out of the process of implementing it. What incentivized me in this case was working more in-depth with the client.txt. My initial idea was to just import the generated guides as plain text in order to feed the overlay, then just work things out from there. Then miscommunication started (I think), but I didn't quite realize it because I was distracted by getting to work with JSON data for the first time ("hey, I'm getting even more out of working on this feature").

I didn't quite understand you expected me to generate the build from the export data (the same way you do in the web version), so that's why I tried -- in my confusion -- parsing it to stitch the data together to form the guide. You asked me if I needed the data in a different format, but I declined out of politeness because of the apparent misunderstanding that I could use this data the same way as a plain-text export. I wouldn't mind implementing it your way, but what irks me is that it's additional work I had not accounted for initially (dynamic overlays, progression system, adapting/restructuring the guides so the client.txt-data trigger advancing the flow, etc.). From my point of view, it's also redundant work and the payoff doesn't really justify it. Making things even worse, I generally have a very hard time working on things that I didn't conceptualize myself, and in this case everything became less and less fun by the minute and more like a chore -- and LLK-UI is a project I persue entirely for fun (it has become a complement to regular gaming).

Please don't understand this as an ultimatum, but I'm putting this feature on hold (for now) because I only accepted the request based on the effort/payoff ratio of having the guides as plain text. There are overlay tools that specialize on leveling guides, so working with the client.txt-data was the only reason I considered implementing a feature like this into my project. Thanks for all your help and being open about cooperating.

**Reply by HeartofPhos** [reply]
No worries, your concerns are similar to the ones I had when I started my project and had to decided between web vs "automatic" tracking, mainly being I couldn't think of a nice way to automate most (if any) state transitions in the route, felt hard to justify another app to download just for next/prev step hotkeys.

My initial reasoning for giving you the raw exported data was giving you as much data as possible that you might be able to link up to client.txt, but from initial understanding and your comments it appears there is indeed limited information to work with, so at best steps could be grouped per area, which would still be awkward for zones which are revisited (e.g. every town area), so doing a text export (grouped or otherwise) would likely end up requiring a whole lot of user input anyway

Appreciate you looking into the feature request, but as you've mentioned it appears to be fairly impractical to implement and other tools already fill the more specific niches

**Reply by Lailloken** [reply]
> My initial reasoning for giving you the raw exported data was giving you as much data as possible

I appreciate it very much. The good old 'making things complicated by trying to help as much as possible' -- I know it too well.

> which would still be awkward for zones which are revisited (e.g. every town area), so doing a text export (grouped or otherwise) would likely end up requiring a whole lot of user input anyway

Not so sure about that. I was planning on parsing the routes and letting the script create steps dynamically. For example, act 3 (steps 62 - 74) with [Enki's arc witch](https://pastebin.com/iSd40YuN) loaded:

```
waypoint -> the library
```
```
hand in a fixture of fate
take lightning penetration support
buy inspiration, onslaught, molten shell, inc duration, determination
buy and omni-click defiance banner
```
```
waypoint -> imperial gardens
```
```
head north and complete the lab-trial
get crafting: all attributes (rank 1)
logout -> the sarn encampment
```
It would just have to make sure there's always a step that can be tracked via client.txt or omni-key at the end of a group (as confirmation it's completed). 'Omni-key' in step 2 refers to a context-sensitive hotkey that I implemented to trigger many specific things. In the event a step is accidentally triggered as complete because of an unexpected disconnect/relog, there'll be some way to go one step back. I also need to double-check which events are logged in the client.txt, maybe there's more than I initially thought.

Step 2 looks a bit bloated, but I could implement a one-click solution for highlighting the required gems in the vendor window (I already use that for a Gwennen feature), so that step would just say something like 'omni-click the vendor search for highlighting, then defiance banner' instead of listing all gems individually. It's hard to say right now if there'll be edge-cases that are difficult to group, but I would just try to fix them when I come across them.

So it's all up to how much work it is to implement plain-text exporting. Doesn't have to be pretty, it can be with or without icon-references (it should include the eight directions, though). If it's too much work, please tell me. I don't want to create unnecessary work for you just to implement a feature (that's more for testing and a small target-group) into my tool -- that wouldn't be worth it. The thing with the arrow sign not rendering properly in AHK happens when the export is loaded from a file -- it's fine when loaded from clipboard.

**Reply by vagrant-soul** [reply]
I'm a beginner. I found an AHK with similar functions before. Maybe you can refer to it
https://github.com/JusKillmeQik/PoE-Leveling-Guide

**Reply by Lailloken** [reply]
> I'm a beginner. I found an AHK with similar functions before. Maybe you can refer to it https://github.com/JusKillmeQik/PoE-Leveling-Guide

Yeah, that's what I was referring to. HeartofPhos and I have been talking outside this thread (it has become too long) and there's an initial test-version up and running. Tracking is working reliably, it has some QoL for buying gems, and I have an idea for an additional feature concerning twink-leveling. Still figuring out if I'll add icons/images or leave the guide as plain-text. v1.28.0 should come out after the weekend.

**Reply by vagrant-soul** [reply]
Does your program support changing the language? I am using Chinese, I can try to modify it to local language if it can be supported. Of course, with your permission

**Reply by Lailloken** [reply]
> Does your program support changing the language? I am using Chinese, I can try to modify it to local language if it can be supported. Of course, with your permission

This is the first time someone requests that. I could restructure the UI to load text-strings from a file instead of being hard-coded into the script. This way, translating everything would be easier and would not require changing the script itself. Since the new league looks simple and probably doesn't require any QoL feature, I might be able to implement that before league launch, but I'll have to see if I find the time.

**Reply by vagrant-soul** [reply]
Thank you very much and look forward to your realization of this function

**Reply by Lailloken** [reply]
> Does your program support changing the language? I am using Chinese, I can try to modify it to local language if it can be supported. Of course, with your permission

Unfortunately, I will have to postpone adding language support for now. I underestimated how much restructuring this requires, and I've been planning on re-doing the UI for some time now, too (a lot of the code is inefficient because it's so old, and some areas don't look nice). So it makes much more sense to add language support when I also re-do the whole UI from the ground up.

If you want to start a Chinese fork anyway, feel free to do so. I don't really have a fixed timeline for the UI overhaul, so it may be a while until then. I don't know how much work it would be to maintain a Chinese fork alongside the main project, and some of my updates change a lot of code at once.

**Comment by Lailloken** [comment]
The struggle has come to an end. Exile-leveling guides can now be imported into LLK-UI to create dynamic overlays, and I threw in some extra features as well. Feel free to try out v1.28.0
